### PR TITLE
Added queryExpiration option to Redis query cache.

### LINKED ISSUE #3586: Add ability to set TTL TimeSpan on redis query storage
**Is your feature request related to a problem? Please describe.**
It would be great if it was possible to set a time to live (TTL) on persisted queries added to redis.

**Describe the solution you'd like**
Add an overload to `.AddRedisQueryStorage` that accepts a `TimeSpan`

### PR DESCRIPTION
**Pull request by gkfischer**
added queryExpiryTimeSpan to AddRedisQueryStorage

- added to RedisQueryStorage constructor
- used in WriteQueryAsync
- Integration tests amended: I know that Snapshots are used in tests all over, but it kinda feels strange to rely on the snapshot result only in the error case. So I threw in an Assert.

Closes #3586

### CODE REVIEW

File: src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/Extensions/HotChocolateRedisPersistedQueriesServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -23,9 +23,12 @@ public static class HotChocolateRedisPersistedQueriesServiceCollectionExtensions
         /// A factory that resolves the redis database that
         /// shall be used for persistence.
         /// </param>
+        /// <param name="queryExpiryTimeSpan">
+        /// A timespan after that a query will be removed from the cache.
+        /// </param>
         public static IServiceCollection AddRedisQueryStorage(
             this IServiceCollection services,
-            Func<IServiceProvider, IDatabase> databaseFactory)
+            Func<IServiceProvider, IDatabase> databaseFactory, TimeSpan? queryExpiryTimeSpan = null)
```
Comment (michaelstaib) [inline src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/Extensions/HotChocolateRedisPersistedQueriesServiceCollectionExtensions.cs:31]: Can you do a line break here to align with the other arguments?

File: src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/Extensions/HotChocolateRedisPersistedQueriesServiceCollectionExtensions.cs [hunk 2]
```diff
@@ -38,7 +41,7 @@ public static class HotChocolateRedisPersistedQueriesServiceCollectionExtensions
             }
 
             return services
-                .AddReadOnlyRedisQueryStorage(databaseFactory)
+                .AddReadOnlyRedisQueryStorage(databaseFactory, queryExpiryTimeSpan)
```
Comment (michaelstaib) [inline src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/Extensions/HotChocolateRedisPersistedQueriesServiceCollectionExtensions.cs:44]: shall we call this queryExpiration?

File: src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/Extensions/HotChocolateRedisPersistedQueriesServiceCollectionExtensions.cs [hunk 3]
```diff
@@ -52,9 +55,12 @@ public static class HotChocolateRedisPersistedQueriesServiceCollectionExtensions
         /// A factory that resolves the redis database that
         /// shall be used for persistence.
         /// </param>
+        /// <param name="queryExpiryTimeSpan">
+        /// A timespan after that a query will be removed from the cache.
+        /// </param>
         public static IServiceCollection AddReadOnlyRedisQueryStorage(
             this IServiceCollection services,
-            Func<IServiceProvider, IDatabase> databaseFactory)
+            Func<IServiceProvider, IDatabase> databaseFactory, TimeSpan? queryExpiryTimeSpan = null)
```
Comment (michaelstaib) [inline src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/Extensions/HotChocolateRedisPersistedQueriesServiceCollectionExtensions.cs:63]: Can you do a line break here to align with the other arguments?

File: src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/Extensions/HotChocolateRedisPersistedQueriesServiceCollectionExtensions.cs [hunk 4]
```diff
@@ -69,7 +75,7 @@ public static class HotChocolateRedisPersistedQueriesServiceCollectionExtensions
             return services
                 .RemoveService<IReadStoredQueries>()
                 .RemoveService<IWriteStoredQueries>()
-                .AddSingleton(sp => new RedisQueryStorage(databaseFactory(sp)))
+                .AddSingleton(sp => new RedisQueryStorage(databaseFactory(sp), queryExpiryTimeSpan))
```
Comment (michaelstaib) [inline src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/Extensions/HotChocolateRedisPersistedQueriesServiceCollectionExtensions.cs:78]: queryExpiration

File: src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/RedisQueryStorage.cs [hunk 5]
```diff
@@ -17,15 +17,20 @@ public class RedisQueryStorage
         , IWriteStoredQueries
     {
         private readonly IDatabase _database;
+        private readonly TimeSpan? _queryExpiryTimeSpan;
```
Comment (michaelstaib) [inline src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/RedisQueryStorage.cs:20]: queryExpiration

File: src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/Extensions/HotChocolateRedisPersistedQueriesServiceCollectionExtensions.cs (hunk 1 above)
Comment (gkfischer) [reply src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/Extensions/HotChocolateRedisPersistedQueriesServiceCollectionExtensions.cs]: sure, haven't looked closely

File: src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/Extensions/HotChocolateRedisPersistedQueriesServiceCollectionExtensions.cs (hunk 4 above)
Comment (gkfischer) [reply src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/Extensions/HotChocolateRedisPersistedQueriesServiceCollectionExtensions.cs]: dunno why I ended up with the type in the name anyway

File: src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/Extensions/HotChocolateRedisPersistedQueriesServiceCollectionExtensions.cs [hunk 6]
```diff
@@ -52,9 +56,13 @@ public static class HotChocolateRedisPersistedQueriesServiceCollectionExtensions
         /// A factory that resolves the redis database that
         /// shall be used for persistence.
         /// </param>
+        /// <param name="queryExpiration">
+        /// A timespan after that a query will be removed from the cache.
```
Comment (michaelstaib) [inline src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/Extensions/HotChocolateRedisPersistedQueriesServiceCollectionExtensions.cs:60]:
```suggestion
        /// A timeout after which a query is removed from the Redis cache.

File: src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/RedisQueryStorage.cs [hunk 7]
```diff
@@ -17,15 +17,20 @@ public class RedisQueryStorage
...
         /// </summary>
         /// <param name="database">The redis database instance.</param>
-        public RedisQueryStorage(IDatabase database)
+        /// <param name="queryExpiration">
+        /// A timespan after that a query will be removed from the cache.
+        /// </param>
+        public RedisQueryStorage(IDatabase database, TimeSpan? queryExpiration = null)
         {
             _database = database
                 ?? throw new ArgumentNullException(nameof(database));
```
Comment (michaelstaib) [inline src/HotChocolate/PersistedQueries/src/PersistedQueries.Redis/RedisQueryStorage.cs:32]:
```suggestion
            _database = database ?? throw new ArgumentNullException(nameof(database));
```

**Review by michaelstaib — APPROVED** [review]

**Comment by michaelstaib** [comment]
- I have removed the query expiration option from the read-only cache variant since it makes no sense there. The read-only query cache is for ahead of time persisted queries and if they would expire clients would stop working. Removing queries from this variant of cache works differently.

- I also added the expiration to the request builder extensions since that is the main way of setting up the cache.

- Lastly, I have reworked the test a bit to include the request builder extension and also retrieve the write cache from the executor builder instead of creating a second instance.

**Pull request merged**